+nosnoc/+dcs/Heaviside.m

```diff
@@ -0,0 +1,229 @@
+
+        f_x
+        
+        dims
+
+        g_lp_stationarity_fun
+        lambda00_fun
+    end
+
+    methods
+        function obj = Heaviside(model)
+            obj.model = model;
+            obj.dims = model.dims;
+        end
+        
+        function generate_variables(obj, opts)
+            switch class(obj.model)
+              case "nosnoc.model.Pss"
+                obj.generate_variables_from_pss(opts);
+              case "nosnoc.model.Heaviside"
+                obj.generate_variables_from_heaviside(opts);
+              otherwise
+                error('A heaviside DCS can only be created from a PSS or Heaviside step model.')
+            end
+        end
+
+        function generate_equations(obj, opts)
+            switch class(obj.model)
+              case "nosnoc.model.Pss"
+                obj.generate_equations_from_pss(opts);
+              case "nosnoc.model.Heaviside"
+                obj.generate_equations_from_heaviside(opts);
+              otherwise
+                error('A heaviside DCS can only be created from a PSS or Heaviside step model.')
+            end
+        end
+    end
+
+    methods(Access=private)
+        function generate_variables_from_pss(obj, opts)
+            import casadi.*
+            dims = obj.dims;
+            model = obj.model;
+
+            % Check pss popluated S and c;
+            if isempty(model.S)
+                error("The PSS model must containt the lp_stationarity matrix S and lp_stationarity function c.");
+            end
+                
+            % dimensions
+            dims.n_alpha = sum(obj.dims.n_c_sys); % number of modes
+            dims.n_lambda = dims.n_alpha;
+            idx = 1;
+            for ii = 1:dims.n_sys
+                sys_idx{ii} = idx:(idx + obj.dims.n_c_sys(ii)-1);
+                idx = idx + obj.dims.n_c_sys(ii);
+                ii_str = num2str(ii);
+                % define alpha (selection of a set valued step function)
+                obj.alpha_sys{ii} = define_casadi_symbolic(opts.casadi_symbolic_mode,['alpha_' ii_str],obj.dims.n_c_sys(ii));
+                obj.alpha = [obj.alpha;obj.alpha_sys{ii}];
+                % define lambda_n_i (Lagrange multipler of alpha >= 0;)
+                obj.lambda_n_sys{ii} = define_casadi_symbolic(opts.casadi_symbolic_mode,['lambda_n_' ii_str],obj.dims.n_c_sys(ii));
+                obj.lambda_n = [obj.lambda_n;obj.lambda_n_sys{ii}];
+                % define lambda_p_i (Lagrange multipler of alpha <= 1;)
+                obj.lambda_p_sys{ii} = define_casadi_symbolic(opts.casadi_symbolic_mode,['lambda_p_' ii_str],obj.dims.n_c_sys(ii));
+                obj.lambda_p = [obj.lambda_p;obj.lambda_p_sys{ii}];
+            end
+
+            % symbolic variables z = [alpha;obj.lambda_n;obj.lambda_p];
+            obj.z_all = [obj.alpha;obj.lambda_n;obj.lambda_p;obj.model.z];
+            obj.dims = dims;
+        end
+
+        function generate_variables_from_heaviside(obj, opts)
```

Comment:
the functions _pss and _heaviside (later also for equaitons)  share a lot of code (Except for the theta and f_x parts, they are almost indentical). 
What speak against of having one function for vars and eqs, and have a if-else or switch-case?

Reply:
No extremely strong opinions here. I think in this case we can absolutely have an if statement. The only thing is sometimes it is nice to have a pseudo-linear way to read the code even if it requires some repetition. Perhaps this is an over-correction on my part however, as I have been quite frustrated during this rewrite with the interleaving `if`s in the old design and disentangling them one by one.

Reply:
I agree, but 90% is repetition and adds quite a few extra lines. Well, most of the old ifs are mostly resolved by having extra classes for every problem type. 

I can live with the repetition, but shorter and still understandable code is also not bad.